hoss/dimension_utilities.py

```diff
@@ -101,10 +101,214 @@ def prefetch_dimension_variables(
+    """This method returns coordinate variables that are referenced
+    in the variables requested. It returns it in a specific order
+    [latitude, longitude]
+    """
+
+    # try:
+    coordinate_variables_set = varinfo.get_references_for_attribute(
+        requested_variables, 'coordinates'
+    )
+    coordinate_variables = []
+    for coordinate in coordinate_variables_set:
+        if varinfo.get_variable(coordinate).is_latitude():
+            coordinate_variables.insert(0, coordinate)
+        elif varinfo.get_variable(coordinate).is_longitude():
+            coordinate_variables.insert(1, coordinate)
+
+    return coordinate_variables
+    # except AttributeError:
+    #    return set()
+
+
+def update_dimension_variables(
+    prefetch_dataset: Dataset,
+    coordinates: Set[str],
+    varinfo: VarInfoFromDmr,
+) -> Dict[str, ndarray]:
+    """Generate artificial 1D dimensions variable for each
+    2D dimension or coordinate variable
+
+    For each dimension variable:
+    (1) Check if the dimension variable is 1D.
+    (2) If it is not 1D and is 2D get the dimension sizes
+    (3) Get the corner points from the coordinate variables
+    (4) Get the x-y max-min values
+    (5) Generate the x-y dimscale array and return to the calling method
+
+    """
+    for coordinate_name in coordinates:
+        coordinate_variable = varinfo.get_variable(coordinate_name)
+        if prefetch_dataset[coordinate_variable.full_name_path][:].ndim > 1:
+            col_size = prefetch_dataset[coordinate_variable.full_name_path][:].shape[0]
+            row_size = prefetch_dataset[coordinate_variable.full_name_path][:].shape[1]
+        crs = get_variable_crs(coordinate_name, varinfo)
+
+    geo_grid_corners = get_geo_grid_corners(prefetch_dataset, coordinates, varinfo)
+
+    x_y_extents = get_x_y_extents_from_geographic_points(geo_grid_corners, crs)
+
+    # get grid size and resolution
+    x_min = x_y_extents['x_min']
+    x_max = x_y_extents['x_max']
+    y_min = x_y_extents['y_min']
+    y_max = x_y_extents['y_max']
+    x_resolution = (x_max - x_min) / row_size
+    y_resolution = (y_max - y_min) / col_size
+
+    # create the xy dim scales
+    x_dim = np.arange(x_min, x_max, x_resolution)
+
+    # The origin is usually the top left and Y values are in decreasing order.
+    if is_latitude_ascending(prefetch_dataset, coordinates, varinfo):
+        y_dim = np.arange(y_max, y_min, y_resolution)
+    else:
+        y_dim = np.arange(y_max, y_min, -y_resolution)
+
+    return {'projected_y': y_dim, 'projected_x': x_dim}
+
+
+def is_latitude_ascending(
+    prefetch_dataset: Dataset,
+    coordinates: Set[str],
+    varinfo: VarInfoFromDmr,
+) -> bool:
+    """
+    Checks if the latitude cooordinate datasets have values
+    that are ascending
+    """
+    lat_arr, lon_arr = get_lat_lon_arrays(prefetch_dataset, coordinates, varinfo)
+    lat_col = lat_arr[:, 0]
+    return is_dimension_ascending(lat_col)
+
+
+def get_lat_lon_arrays(
```

Comment:
There is heavy overlap between this function and `get_coordinate_variables`. Perhaps either one of the functions is unnecessary, or maybe this could be reworked to call `get_coordinate_variables` and then return the arrays from the output of that, instead of duplicating the meat of the logic.

Reply:
I think the get_coordinate_variables just gets the list of coordinates. The list is used by the get_lat_lon_arrays to get the lat/lon arrays. Maybe your comment was for an earlier version? Let me know if I still need to rework this.

Reply:
You're right, sorry, the return value is a list of coordinate names, but the idea is still applicable. (The main looping logic is essentially duplicated between the two functions, see L168-L174 versus L278-L283: those are doing the same loops and the same checks)

That said, I think there's a bigger concern here for collections with multiple grids (and therefore multiple groups of variables in the `coordinates` references), such as SPL3FTP. In the function below the logic is going to go through all `coordinates` and check each one to see if it is a latitude or a longitude. You're getting the list of coordinates (via a few other functions) from `get_coordinate_variables` (originally called [here](https://github.com/nasa/harmony-opendap-subsetter/pull/15/files#diff-d78248962545c9922cc4c65404061f867387f035a2e5cbb92fb5776a9f3688d5R126)), which would have multiple latitude/longitude references for SPL3FTP (if a user requests both polar and global variables). If you are trying to handle both the global and polar grid, you're going to need to sometimes get the global coordinates and sometimes the polar, but this function will always just get whichever are listed last (because the loop will get to them last and assign them to `lat_arr` and `lon_arr`, even if something is already assigned to those variables).

Reply:
Similar duplication on L381 - L386.

Reply:
Thanks Owen. I see that... - updated. - https://github.com/nasa/harmony-opendap-subsetter/pull/15/commits/802fe0ec04b853a7a7abb4718e2a05b0e8e6ed3c